src/qnx/qinjector-glue.c

```diff
@@ -0,0 +1,738 @@
+{
+  gboolean success = FALSE;
+  gint ret;
+  const gchar * failed_operation;
+  gint fd;
+  gint i;
+  gchar as_path[PATH_MAX];
+  pthread_t tid;
+  debug_thread_t thread;
+  procfs_greg saved_registers, modified_registers;
+  procfs_status status;
+  procfs_run run;
+  sigset_t * run_fault = (sigset_t *) &run.fault;
+
+  sprintf(as_path, "/proc/%d/as", pid);
+  fd = open (as_path, O_RDWR);
+  g_assert (fd != -1);
+
+  /*
+   * Find the first active thread:
+   */
+  for (tid = 1;; tid++)
+  {
+    thread.tid = tid;
+    if (devctl (fd, DCMD_PROC_TIDSTATUS, &thread, sizeof (thread), 0) == EOK)
+      break;
+  }
+
+  /*
+   * Set current thread and freeze our target thread:
+   */
+  ret = devctl (fd, DCMD_PROC_CURTHREAD, &tid, sizeof (tid), 0);
+  CHECK_OS_RESULT (ret, ==, EOK, "DCMD_PROC_CURTHREAD");
+
+  ret = devctl (fd, DCMD_PROC_STOP, &status, sizeof (status), 0);
+  CHECK_OS_RESULT (ret, ==, EOK, "DCMD_PROC_STOP");
+
+  if (status.state == STATE_DEAD)
+    goto beach;
+
+  if (status.state != STATE_STOPPED)
+  {
+    /*
+     * If the thread was not in the STOPPED state, it's probably
+     * blocked in a NANOSLEEP or some syscall. We'll SIGHUP
+     * it to kick it out of the blocker and WAITSTOP until the
+     * signal is delivered.
+     */
+    memset(&run, 0, sizeof (run));
+    run.flags |= _DEBUG_RUN_TRACE;
+    sigemptyset (&run.trace);
+    sigaddset (&run.trace, SIGHUP);
+    ret = devctl (fd, DCMD_PROC_RUN, &run, sizeof (run), 0);
+    CHECK_OS_RESULT (ret, ==, EOK, "DCMD_PROC_RUN");
+
+    kill (pid, SIGHUP);
+
+    ret = devctl (fd, DCMD_PROC_WAITSTOP, &status, sizeof (status), 0);
+    CHECK_OS_RESULT (ret, ==, EOK, "DCMD_PROC_WAITSTOP");
+    if (status.why == _DEBUG_WHY_TERMINATED)
+      goto beach;
+
+    /*
+     * We need the extra PROC_STOP because status.state is not
+     * properly reported by WAITSTOP.
+     */
+    ret = devctl (fd, DCMD_PROC_STOP, &status, sizeof (status), 0);
+    CHECK_OS_RESULT (ret, ==, EOK, "DCMD_PROC_STOP");
+  }
+
+  /*
+   * Get the thread's registers:
+   */
+  ret = devctl (fd, DCMD_PROC_GETGREG, &saved_registers, sizeof (saved_registers), 0);
+  CHECK_OS_RESULT (ret, ==, EOK, "DCMD_PROC_GETGREG");
+
+  memcpy(&modified_registers, &saved_registers, sizeof (saved_registers));
+
+  /*
+   * Set the PC to be the function address and SP to the stack address.
+   */
+  if ((func & 1) != 0)
+  {
+    modified_registers.arm.gpr[ARM_REG_PC] = (func & ~1);
+    modified_registers.arm.spsr |= PSR_T_BIT;
+  }
+  else
+  {
+    modified_registers.arm.gpr[ARM_REG_PC] = func;
+    modified_registers.arm.spsr &= ~PSR_T_BIT;
+  }
+
+  for (i = 0; i < args_length && i < 4; i++)
+  {
+    modified_registers.arm.gpr[i] = args[i];
+  }
+
+  for (i = args_length - 1; i >= 4; i--)
+  {
+    modified_registers.arm.gpr[ARM_REG_SP] -= 4;
+
+    if (!frida_remote_write_fd (fd, modified_registers.arm.gpr[ARM_REG_SP], &args[i],
+          4, error))
+    {
+      close (fd);
+      return FALSE;
```

Comment:
Hmm, it's generally not a good idea to have arbitrary return points in functions with unwind logic (goto beach).